OpenOversight/migrations/versions/562bd5f1bc1f_add_order_to_jobs.py

```diff
@@ -0,0 +1,40 @@
+"""Add order to jobs in DB and adjust constraints on order column
+
+Revision ID: 562bd5f1bc1f
+Revises: 6045f42587ec
+Create Date: 2020-04-24 01:58:05.146902
+
+"""
+from alembic import op
+import sqlalchemy as sa
+from app.models import Department, Job, db
+
+
+# revision identifiers, used by Alembic.
+revision = '562bd5f1bc1f'
+down_revision = '6045f42587ec'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.drop_constraint('unique_department_job_order', 'jobs', type_='unique')
+
+    for department in Department.query.all():
+        job_order = 0
+        for job in Job.query.filter_by(department_id=department.id):
```

Comment:
this looks great! one thought: here should we also filter by those jobs that don't have a value for `order` yet? (in case someone running this migration on their data already did define `order` for jobs for a department)

Reply:
That makes sense! :+1: I will add that filter.

Reply:
Hey @redshiftzero, I'm having a hard time filtering for jobs where `order` is null.  I tried to write a query using `filter` instead of `filter_by`.  I tested it out in `test_utils.py` with `test_find_jobs_without_order`, and it seemed to work.

But, when I run `flask db upgrade`, it hangs after this line:

`INFO  [alembic.runtime.migration] Running upgrade 6045f42587ec -> 562bd5f1bc1f, Add order to jobs in DB and adjust constraints on order column` 

I tried to see if there are any locks on the database

`SELECT (locktype, pid, mode, granted) FROM pg_locks ORDER BY pid;`

and found this process where `granted` was false 

` (relation,83,AccessShareLock,f)`


I checked `pg_stat_activity` next

`SELECT (pid, application_name, state, query) FROM pg_stat_activity WHERE pid=83;`
` (83,"",active,"SELECT jobs.id AS jobs_id, jobs.job_title AS jobs_job_title, jobs.is_sworn_officer AS jobs_is_sworn_officer, jobs.""order"" AS jobs_order, jobs.department_id AS jobs_department_id +
 FROM jobs                                                                                                                                                                                          +
 WHERE jobs.department_id = 1 AND jobs.""order"" IS NULL")`




It seems to be locked on this query?  I'm not sure why. 


Then I investigated another pid that appeared in the `pg_locks` table.  It seems that the process to remove the unique constraint on job order is idle?

`SELECT (pid, application_name, state, query) FROM pg_stat_activity WHERE pid=82;`
 `(82,"","idle in transaction","ALTER TABLE jobs DROP CONSTRAINT unique_department_job_order")`


I'm having a hard time figuring out why the migration isn't completing.  Any help would be appreciated.  Thanks!

Reply:
Update: after much investigating and duckduckgoing, I figured out that the migration was hanging due to my use of `db.session.commit()`.  I wrote a new migration that uses `connection` instead of `session` and populates the order column for unordered ranks with 0. 

I was able to successfully run this on a mockdata department that had unordered ranks, then locally view the ranks and update them on the edit department page.  However, when I tried to run `make test`, it failed, because the department with unordered ranks broke the new not-null constraint on department order.  Consequently, I added rank order to the department in order to make the tests pass.

Let me know if you have any questions.